lib/elixir/lib/calendar/datetime.ex

```diff
@@ -172,20 +172,213 @@ defmodule DateTime do
+  date and time in a certain time zone.
+
+      iex> {:ok, datetime} = DateTime.from_naive(~N[2018-07-28 12:30:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> datetime
+      #DateTime<2018-07-28 12:30:00+02:00 CEST Europe/Copenhagen>
+
   """
   @doc since: "1.4.0"
-  @spec from_naive(NaiveDateTime.t(), Calendar.time_zone()) :: {:ok, t}
-  def from_naive(naive_datetime, time_zone)
+  @spec from_naive(
+          NaiveDateTime.t(),
+          Calendar.time_zone(),
+          TimeZoneDatabaseClient.tz_db_or_config()
+        ) ::
+          {:ok, t}
+          | {:outside_leap_second_data_range, t}
+          | {:ambiguous, t, t}
+          | {:gap, t, t}
+          | {:error, :time_zone_not_found}
+          | {:error, :incompatible_calendars}
+          | {:error, :no_time_zone_database}
+
+  def from_naive(naive_datetime, time_zone, tz_db_or_config \\ :from_config)
+
+  def from_naive(%{second: 60} = naive_datetime, "Etc/UTC", tz_db_or_config) do
+    {:ok, dt} = do_from_naive(naive_datetime, "Etc/UTC", 0, 0, "UTC")
+
+    case validate_positive_leap_second(dt, tz_db_or_config) do
+      :ok ->
+        {:ok, dt}
+
+      {:error, :outside_leap_second_data_range} ->
+        {:outside_leap_second_data_range, dt}
+
+      error ->
+        error
+    end
+  end
+
+  def from_naive(naive_datetime, "Etc/UTC", _) do
+    do_from_naive(naive_datetime, "Etc/UTC", 0, 0, "UTC")
+  end
+
+  def from_naive(%{calendar: Calendar.ISO} = naive_datetime, time_zone, tz_db_or_config) do
+    case TimeZoneDatabaseClient.time_zone_periods_from_wall_datetime(
+           naive_datetime,
+           time_zone,
+           tz_db_or_config
+         ) do
+      {:single, period} ->
+        do_from_naive_check_leap_second(
+          naive_datetime,
+          time_zone,
+          period.std_offset,
+          period.utc_offset,
+          period.zone_abbr,
+          tz_db_or_config
+        )
+
+      {:ambiguous, first_period, second_period} ->
+        {:ok, first_datetime} =
+          do_from_naive(
+            naive_datetime,
+            time_zone,
+            first_period.std_offset,
+            first_period.utc_offset,
+            first_period.zone_abbr
+          )
+
+        {:ok, second_datetime} =
+          do_from_naive(
+            naive_datetime,
+            time_zone,
+            second_period.std_offset,
+            second_period.utc_offset,
+            second_period.zone_abbr
+          )
+
+        {:ambiguous, first_datetime, second_datetime}
+
+      {:gap, {first_period, first_period_until_wall}, {second_period, second_period_from_wall}} ->
+        # `until_wall` is not valid, but any time just before is.
+        # So by subtracting a second and adding .999999 seconds
+        # we get the last microsecond just before.
+        before_naive =
+          first_period_until_wall
+          |> Map.put(:microsecond, {999_999, 6})
+          |> NaiveDateTime.add(-1)
```

Comment:
@lau why can't we use first_period_until_wall as is and then document that the `:gap` function returns the first invalid value and the first valid value from then?

Reply:
We can't do what I propose because then it would return an invalid datetime. I wonder then if we should ask the timezone database to return the last valid instant so we can skip this conversion. Thoughts @lau?

Reply:
Moved this to the new issue.